### PR TITLE
Competitive programming

### PR DESCRIPTION
**Pull request by iam-sakshi**
<hr>

## Description 📜
Summary : In this documentation , Binary Search simple applications are demonstrated with the aim that it can help others to look for binary search in competitive programming for other uses too . In this documentation , the requirement of Binary Search is discussed and programs like:
- first occurrence
- last occurrence
- Count of the element 
- floor of the element
- Ceil of the element
the programming language used in this is C++.
issue name :Competitive Programming : Binary Search  
issue number : #7131

<hr>

## Type of change 📝

- [ ] Au#dio (Should be in mp3 format Includes speech clarity, Concise ,Low distortion)
- [ ] Vi#deo (Animations, screen-recordings, presentations and regular explanatory films are all possibilties etc)
- [x] Doc#umentation (Content Creation in the form of codes or tutorials)
- [ ] Other (If you choose other, Please mention changes below) 

<hr>

## Domain of Contribution 📊

- [ ] Android Dev #(Flutter)
- [ ] Android Dev #(Java)
- [ ] Android Dev #(Kotlin)
- [ ] Backend Dev #(Java)
- [ ] Backend Dev #(.NET)
- [ ] Backend Dev #(PHP)
- [ ] Backend Dev #(Python)
- [ ] C/#CPP
- [x] Competitive #Programming
- [ ] Cyber #Security
- [ ] DS#A
- [ ] Data#base
- [ ] Datascience with #Python
- [ ] Datascience with #R
- [ ] Frontend Dev #HTML/CSS/JS
- [ ] Frontend Dev #React/Angular/Vue
- [ ] Go#lang
- [ ] Interview #Prep
- [ ] Java_#Domain
- [ ] JavaScript_#Domain
- [ ] ME#RN
- [ ] Machine #Learning
- [ ] Open #Source
- [ ] Python_#Domain
- [ ] Ru#st
- [ ] Statis#tics
- [ ] UI/#UX

<hr>

## Checklist ✅

- [x] I follow [Contributing Guidelines](https://github.com/girlscript/winter-of-contributing/blob/main/.github/CONTRIBUTING.md) & [Code of conduct](https://github.com/girlscript/winter-of-contributing/blob/main/.github/CODE_OF_CONDUCT.md) of this project.
- [x] I have performed a self-review of my own code or work.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] My changes generates no new warnings.
- [x] I'm GWOC'21 contributor

<hr>

## Screenshots / Gif (Optional) 📸

<hr>

### CODE REVIEW

**Comment by iam-sakshi** [comment]
please review my PR

**Review by ravireddy07 — CHANGES_REQUESTED** [review]
Please update the following things:

* remove spaces in filename, rename it to like `BinarySearch.md`.
* Can you explain the more on each sub-topic, like what exactly floor/ceil in this binary search doc and how it's related.
* add this explanation on top and code below. Like: Title->explanation->code...
* Add complexity's to each approach.

**Comment by iam-sakshi** [comment]
okay

File: Competitive_Programming/C++/Searching & Sorting/Binary Search/BinarySearchApplication.md [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+```
+>Output:
+>
+>Floor of the element 3
+>
+>Ceil of the element 5
+
+>Time Complexity O(log n)    , where n is the size of array
+
+Explanation:
```
Comment (ravireddy07) [inline Competitive_Programming/C++/Searching & Sorting/Binary Search/BinarySearchApplication.md:182]:
```suggestion
## Approach:
```

File: Competitive_Programming/C++/Searching & Sorting/Binary Search/BinarySearchApplication.md [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+- In floor of an element in the array we need to find the greatest element samaller than the element we wish to find the floor.
+- example:
+ array =[1,2,3,5,6,7,8] element = 4 for which we have to find the floor , 
+ 1. why Binary Search ?  because the array is sorted and we know that the element floor will be present in the array (ie. from start=0 to end=n-1 where n is the size of array)
+2. so we find the mid element using start and end of the array index till the time start <= end.
+3. then we compare array[mid] with the element for which we have to find floor.
+4. and whenever we find any smaller element than 4 ,then we save it in a variable indx(as that can be a possible answer) and move the start to mid+1 , to find if other greater element that is smaller than the element (in this case 4).
+5. After Binary Search, we will return the indx as it will have the floor of the element.
+- Similarily we can approach ceil, as we know ceil is the smallest greater element than the element for which we want to find ceil.
```
Comment (ravireddy07) [inline Competitive_Programming/C++/Searching & Sorting/Binary Search/BinarySearchApplication.md:191]: Place this above the code

File: Competitive_Programming/C++/Searching & Sorting/Binary Search/BinarySearchApplication.md [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+
+These are some of the programs that can be solve by Binary Search , Binary search is really important in competitve programming as it do the searching in O(log (n)) time .
```
Comment (ravireddy07) [inline Competitive_Programming/C++/Searching & Sorting/Binary Search/BinarySearchApplication.md:193]: Add this complexity with title.

Comment (ravireddy07) [inline Competitive_Programming/C++/Searching & Sorting/Binary Search/BinarySearchApplication.md:193]:
```suggestion
## Complexity
These are some of the programs that can be solve by Binary Search , Binary search is really important in competitve programming as it do the searching in ```O(log (n))``` time .
```

**Review by ravireddy07 — CHANGES_REQUESTED** [review]

**Comment by iam-sakshi** [comment]
Done all the suggested changes.

File: Competitive_Programming/C++/Searching & Sorting/Binary Search/BinarySearchApplication.md [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+>Ceil of the element 5
+
+>Time Complexity O(log n)    , where n is the size of array
+
+## Complexity
+These are some of the programs that can be solve by Binary Search , Binary search is really important in competitve programming as it do the searching in ```O(log (n))``` time .
+
+>The flow of document
+>
+>Title -> explanation -> code -> Time Complexity
```
Comment (ravireddy07) [inline Competitive_Programming/C++/Searching & Sorting/Binary Search/BinarySearchApplication.md:199]: remove this

**Review by ravireddy07 — CHANGES_REQUESTED** [review]
please change following things;

* Add formatted code.
* Place the explanation above and the code below.

Hope you got this.

**Comment by iam-sakshi** [comment]
Done

**Review by ravireddy07 — APPROVED** [review]

**Comment by ravireddy07** [comment]
Good work @iam-sakshi  👍

**Comment by ravireddy07** [comment]
@iam-sakshi  Please Update the status of PR in PR Tracksheet

**Pull request merged**